### DISCUSSION TITLE
Tasmota TCP client- not sure if possible

### DISCUSSION THREAD

**Discussion by mumin50** [arendst/Tasmota#16503] [Q&A] [answered]
Hi All
I have specific use case and trying to wrap my head around it.
I have a device that sends out all logs on TCP server on ip192.168.0.10 on port 10000.
That device on start of process logs start always with same string.
I want that tasmota on esp32 triggers on that line and turn a Led strip light for 10s.

Probably i can easy do that on node-red by conecting as tcp client and forward it toesp32 via mqtt, but i think is waste of resources...

Is there any way to conect to TCP server and listen to it till that string is there and trigger on it

**Comment by lalo-uy** [comment]
Looks like an easy task for an Arduino script on the esp32.

**Comment by sfromis** [comment, marked as answer]
On ESP32, Tasmota does already have basic [TCP client](https://tasmota.github.io/docs/Berry/#webserver-module) support in the Berry language, where you can connect and read/write data. That would include looking for a string, and do something if matched.

**Reply by mumin50** [reply]
Thank you i will look in to it

**Comment by s-hadinger** [comment]
I'm missing information here so I need to guess. Tasmota can connect to the TCP server but it's a big assumption that the server will duplicate the traffic received from the other device and send it back to Tasmota

**Reply by sfromis** [reply]
What other device do you infer from what was mentioned? The text looks like a device which itself is providing a TCP server for connecting to. No duplication for that, AFAICT.
> I have a device that sends out all logs on TCP server

**Comment by lalo-uy** [comment]
I infer the device sends TO a TCP server, aka syslog

**Comment by barbudor** [comment]
So you want the device to send it's logs TO Tasmota INSTEAD of the syslog server?
Because of course, tasmota cannot spy on a TCP connection between the device and the syslog server...

**Reply by sfromis** [reply]
If the sender is using the syslog format (not clarified), that could fit with a syslog server being implemented in Tasmota, possibly using the TCP client/server features available with Berry.

**Comment by mumin50** [comment]
Ok made some progress with berry scripting by using tcp client in Berry 

tcp = tcpclient()
tcp.connect("192.168.2.10", 12111)
print("connected:", tcp.connected())
s= "hallo"
tcp.write(s)
print("available1:", tcp.available())
tasmota.delay(2000)
print("available1:", tcp.available())
r = tcp.read()
tcp.close()
print(r)

And i get response from from server 

connected: true
available1: 0
available1: 8
[1003]

where that [1003] is exactly what i need. Now i need to learn how to do some not blocking while loops and i will be ok. 
Can someone help how to recheck every 100ms if tcp.avaliable and if yes than trigger command in tasmota

**Reply by sfromis** [reply]
To wait for things, do not use loops.

If you put your code in a class, add it as a driver using `tasmota.add_driver(self)`, you can have a method named `every_100ms()`, which will be called every 100 ms... Executing Tasmota commands from Berry is `tasmota.cmd(...)`, but a few functions (like power on/off) can also be controlled directly from Berry code.

https://tasmota.github.io/docs/Berry/#creating-a-tasmota-driver

**Reply by sfromis** [reply]
As the `every....` routines should be kept rather fast, it can be useful to delay actual code executing using `tasmota.set_timer` with 0 milliseconds delay, meaning that the code will then be executed next main loop iteration, without risk of it causing trouble by taking some time.

**Reply by mumin50** [reply]
Im getting there... :)
i need to have it this way that berry script opens tcp this way is connected and open and periodicly do tcp.avaliable and tcp.read. 
that i can do with class

How to start it all on esp reboot. i need to make it to .be file? and call it speciall way?
after adding driver it stay there after reset or i need to re-add it somechow...

If that works i can remove so many raspberry pi that i used to make this before... Berry in tasmota has grat future if more examples will folow...

**Reply by sfromis** [reply]
Yes, Berry code needs to be stored in a `.be` file. You can get it started after boot by having the code in `autoexec.be`, or by loading your Berry script file from there.

**Reply by mumin50** [reply]
im getting confused 
i made this 

_tcp = tcpclient()
class MyDriver
 def every_second()
  if tcp.connected == false
   tcp.connect("192.168.2.10", 12111)
   var s= "hallo - client connected"
   tcp.write(s)
   print("connected: ", tcp.connected())
   print("sent hallo...")
  end
  while tcp.available() > 0
   var r = tcp.read()
   print("recived: ", r)	
  end
 end
end_

_d1 = MyDriver()
tasmota.add_driver(d1)_

it should add a driver, but how to use driver... 
for now i know that i need to put this code in autoexec.be that is runs that script on start.... 
it should add a driver, but how to use driver...

**Reply by mumin50** [reply]
let me explain what i need with my code 
first i need to as setup of my script say 

_tcp = tcpclient()_

I need to do that only once(!) as if do it again open connection get closed.
then i need to run every second

_if tcp.connected == false
 tcp.connect("192.168.2.10", 12111)
 var s= "hallo - client connected"
 tcp.write(s)
 print("connected: ", tcp.connected())
 print("sent hallo...")
end
while tcp.available() > 0
 var r = tcp.read()
 print("recived: ", r)
end_

If i do that by hand in berry console works perfectly

**Reply by sfromis** [reply]
The driver is being used by having been added as a Tasmota driver, meaning that if this was done correctly, Tasmota will call your every_second method automatically. You could verify by having something like a print('every_second was called') there

If having the code in `autoexec.be`, do check the console (not the Berry console) for any signs of error messages, scrolling back right after boot. For testing, you can also take the full code from `autoexec.be` and copy/paste into the Berry scripting console.

One small error, instead of `if tcp.connected == false` you should have `if tcp.connected() == false`, or maybe `if ! tcp.connected()`

As a code style recommendation, I'd suggest having the tcp variable inside the class, instead of as a global variable, and have the class add itself:
```
class MyDriver
  var tcp
  def init()
    self.tcp = tcpclient()
    tasmota.add_driver(self)
  end
  def every_second()
    if self.tcp.connected() == false
      self.tcp.connect("192.168.2.10", 12111)
  ...
```
(I also like using indentation for enhanced readability)

**Comment by mumin50** [comment]
this is final code :

```
var tcp = tcpclient()
var ip = "192.168.2.10"
var port = 12111

tcp.connect(ip, port)
class MyDriver
	def every_100ms()
		if tcp.connected() == false
			tcp.connect(ip, port)
			print("Server re-connected: ", tcp.connected())
		end
		while tcp.available() > 0
			var r = tcp.read()
			print(r)	
		end
	end
end

d1 = MyDriver()
tasmota.add_driver(d1)

```
probably don't need to check every 100ms... but not sure how big is buffor...
It is still just a draft for connecting, and now it needs recived data procesing, but it is perfect start. 
Thanks all and specially **sfromis**

**Reply by s-hadinger** [reply]
There is a very low overhead if you check every 100ms, no problem with that.
